File: library/src/main/java/com/pigcms/library/android/adapter/DiffCallback.java
package com.pigcms.library.android.adapter;

import android.support.v7.util.DiffUtil;

import com.pigcms.library.android.entify.BasicVo;

import java.util.List;

/**这是一个rvadapter比较工具，对于不是xrecyclerview封装的 wrapdaadapter可以适用，若出现addheaderview错位，你需要重写AdapterDataObserver调整position 你问为什么不重写Callback
 * Created by win7 on 2017-03-31. 主要针对有结构顺序模块的数据，特别是重新刷新小部分或单个数据更变的情形
 */

public class DiffCallback extends DiffUtil.Callback {
    private List<BasicVo> mOldeDatas,mNewDatas;
    public DiffCallback (List<BasicVo> mOldeDatas,List<BasicVo> mNewDatas){
        this.mNewDatas=mNewDatas;
        this.mOldeDatas=mOldeDatas;
    }

    @Override
    public int getOldListSize() {
        return mOldeDatas==null?0:mOldeDatas.size();
    }

    @Override
    public int getNewListSize() {
        return mNewDatas==null?0:mNewDatas.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        //需要判断是否正确的问题
        return true;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return true;
    }
}

File: app/src/main/java/com/example/sinner/letsteacher/utils/DiffCallback.kt
package com.example.sinner.letsteacher.utils

import android.support.v7.util.DiffUtil
import com.example.sinner.letsteacher.activity.clearboom.BoomItem

/**
 * Created by sinner on 2017-10-28.
 * mail: wo5553435@163.com
 * github: https://github.com/wo5553435
 */
class DiffCallback(val olddatas:List<BoomItem>,val newdatas:List<BoomItem>) :DiffUtil.Callback(){


    override fun getOldListSize()=(olddatas?.size)

    override fun getNewListSize()=(newdatas?.size)

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        var olddata=olddatas[oldItemPosition] ;var newdata=newdatas[newItemPosition];
        return (olddata.isclick==newdata.isclick)&&(olddata.isshow==newdata.isshow)
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int)=
            (olddatas.get(oldItemPosition).x==newdatas.get(newItemPosition).x)&&(olddatas.get(oldItemPosition).y==newdatas.get(newItemPosition).y)

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        if (olddatas[oldItemPosition].isclick==newdatas[newItemPosition].isclick&&olddatas[oldItemPosition].isshow==newdatas[newItemPosition].isshow)
            return null
        return 1
    }
}